src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,232 @@
+
+#nullable enable
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using StreamJsonRpc;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
+{
+    internal class InProcLanguageServer
+    {
+        private readonly IDiagnosticService _diagnosticService;
+        private readonly JsonRpc _jsonRpc;
+        private readonly IThreadingContext _threadingContext;
+        private readonly LanguageServerProtocol _protocol;
+        private readonly Workspace _workspace;
+
+        private VSClientCapabilities? _clientCapabilities;
+
+        public InProcLanguageServer(Stream inputStream, Stream outputStream, LanguageServerProtocol protocol, Workspace workspace, IDiagnosticService diagnosticService, IThreadingContext threadingContext)
+        {
+            this._protocol = protocol;
+            this._workspace = workspace;
+            this._threadingContext = threadingContext;
+
+            this._jsonRpc = new JsonRpc(outputStream, inputStream, this);
+            this._jsonRpc.StartListening();
+
+            this._diagnosticService = diagnosticService;
+            this._diagnosticService.DiagnosticsUpdated += DiagnosticService_DiagnosticsUpdated;
+        }
+
+        /// <summary>
+        /// Handle the LSP initialize request by storing the client capabilities
+        /// and responding with the server capabilities.
+        /// The specification assures that the initialize request is sent only once.
+        /// </summary>
+        [JsonRpcMethod(Methods.InitializeName)]
+        public InitializeResult Initialize(JToken input)
+        {
+            // InitializeParams only references ClientCapabilities, but the VS LSP client
+            // sends additional VS specific capabilities, so directly deserialize them into the VSClientCapabilities
+            // to avoid losing them.
+            this._clientCapabilities = input["capabilities"].ToObject<VSClientCapabilities>();
+
+            return new InitializeResult
+            {
+                Capabilities = new VSServerCapabilities
+                {
+                    DocumentHighlightProvider = true,
+                    DocumentSymbolProvider = true,
+                    DocumentFormattingProvider = true,
+                    DocumentRangeFormattingProvider = true,
+                    DocumentOnTypeFormattingProvider = new DocumentOnTypeFormattingOptions { FirstTriggerCharacter = "}", MoreTriggerCharacter = new[] { ";", "\n" } },
```

Comment:
First, i don't understand this list.  For example, i see this in FormatCommandHandler:

```
        private bool ValidSingleOrMultiCharactersTokenKind(char typedChar, SyntaxKind kind)
            => typedChar switch
            {
                'n' => kind == SyntaxKind.RegionKeyword || kind == SyntaxKind.EndRegionKeyword,
                't' => kind == SyntaxKind.SelectKeyword,
                'e' => kind == SyntaxKind.WhereKeyword,
                _ => true,
            };
```

So it looks like we're missing some characters.

Second, this is a layering violation.  We're in a language-neutral layer, but we're hardcoding in auto-format characters.  What about TS/F#/VB? etc.

Reply:
Yes characters are missing.  This is not changed by this PR, see http://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Initialize/InitializeHandler.cs,f7eb4fee5281789a

This should probably actually just be delegating to the handler one now that I think about it.  Will make that change.  

Filed https://github.com/dotnet/roslyn/issues/40389 to track making the trigger characters correct.

F# (is implementing) and TS (will implement) there own language server, so this needs to handle C# and VB.  Not sure if VB is handled properly, but will followup in https://github.com/dotnet/roslyn/issues/40389

Reply:
Wait...  So f#/TS/js will have to redo all the work Roslyn already does for them? Isn't that a worse place to be?